### ISSUE TITLE
Could you please provide the UMD dataset used for reproduction?

### ISSUE THREAD

**Issue by hiwmy0211** [Reagan1311/OOAL#3]
hello, thank you for your excellent work again! May I ask if you could provide the UMD dataset used for training and testing in the network? After downloading the original UMD dataset, I found that it cannot be directly fed into the network for reproduction. Thank you in advance for your reply!

**Comment by Reagan1311** [comment]
Hi, here is the UMD one-shot data we used for training: [link](https://drive.google.com/file/d/1MM8n9x68nbRs4xf0zJdgnnq6v1w5ad5m/view?usp=sharing). Following AffCorrs, we have changed the affordance of Ladle from "Contain" to "Scoop". For training and testing, we used mmsegmentation to conduct the experiments.

**Comment by hiwmy0211** [comment]
![1](https://github.com/user-attachments/assets/a11a08bd-4140-4729-be49-f41f7537eb87)
Thank you again for your response! However, the labels downloaded from this link are all black. Is there an issue with the link?

**Comment by Reagan1311** [comment]
Hi, this is because the value in the label image is from 0~7 (background + 7 affordances). You can assign each value with different RGB colors to visualize it.

**Comment by hiwmy0211** [comment]
Thank you again for your response. I now have a clearer understanding of the training and testing settings under the UMD dataset for both Seen and Unseen configurations. My understanding is that the test sets for both Seen and Unseen configurations in the UMD dataset are the same, consisting of all the data provided by the official website. Could you please confirm if this understanding is correct?
Additionally, could you provide the network deployed on the UMD dataset? I noticed that the evaluation metrics used are different from those on the AGD20K dataset, so I would like to ask if it would be possible for you to share the code used for running the network on the UMD dataset.
Thank you in advance for your reply!

**Comment by Reagan1311** [comment]
We use the official test sets from UMD dataset, but they are different for seen and unseen settings. 
After you downloading the dataset, you will find the `category_split` file is for seen setting, and the `novel_split` file is for unseen setting.
The network is the same for the UMD experiments, you can try to implement it using the MMSegmentation toolbox. The code I previously used was stored on a server, but I no longer have access to it.

**Comment by hiwmy0211** [comment]
Thank you very much for your response. I will deploy the network running on the UMD dataset based on your suggestions. 
At the same time, I have one more question I'd like to ask: how can networks like ZegCLIP and SAN be applied to the AGD20K dataset? I noticed that networks like ZegCLIP have specific requirements for datasets, particularly regarding the color mapping of different class labels. For example, the "car" class is represented by the color [0, 64, 96]. However, the AGD20K dataset uses sparse annotation, where even different affordance labels are assigned values close to 255. Given this, I would like to ask how ZegCLIP distinguishes between different affordance labels when reproducing the network on the AGD20K dataset. If possible, would it be possible for you to provide the code used for reproduction for reference and study?
Thank you in advance for your reply!

**Comment by Reagan1311** [comment]
The color mapping is specific to each dataset for visualization. Any typical or open-vocabulary segmentation networks can be used for the AGD20K, as long as the raw output is constrained between 0 and 1, as AGD20K is annotated with soft labels in this range.
For experiments, we use the official ZegCLIP code, which applies a Sigmoid function in the end.

**Comment by hiwmy0211** [comment]
Thank you very much for your response. May I ask how to deploy the ZegClip network on the weakly supervised AGD20K dataset during the reproduction process? During training, is it also limited to using only one image per object category?

**Comment by MUVGuan** [comment]
> We use the official test sets from UMD dataset, but they are different for seen and unseen settings. After you downloading the dataset, you will find the `category_split` file is for seen setting, and the `novel_split` file is for unseen setting. The network is the same for the UMD experiments, you can try to implement it using the MMSegmentation toolbox. The code I previously used was stored on a server, but I no longer have access to it.

@Reagan1311 Hello, why do UMD experiments need to be done on MMSegmentation? Can I not use it and directly train according to the training code of AGD20K?

**Comment by Reagan1311** [comment]
> > We use the official test sets from UMD dataset, but they are different for seen and unseen settings. After you downloading the dataset, you will find the `category_split` file is for seen setting, and the `novel_split` file is for unseen setting. The network is the same for the UMD experiments, you can try to implement it using the MMSegmentation toolbox. The code I previously used was stored on a server, but I no longer have access to it.
> 
> [@Reagan1311](https://github.com/Reagan1311) Hello, why do UMD experiments need to be done on MMSegmentation? Can I not use it and directly train according to the training code of AGD20K?

Hi, @MUVGuan. You can do the UMD experiments based on the training code of AGD20K or any other codebase. I used MMseg because it's convenient to implement other segmentation models.

**Comment by MUVGuan** [comment]
> > > We use the official test sets from UMD dataset, but they are different for seen and unseen settings. After you downloading the dataset, you will find the `category_split` file is for seen setting, and the `novel_split` file is for unseen setting. The network is the same for the UMD experiments, you can try to implement it using the MMSegmentation toolbox. The code I previously used was stored on a server, but I no longer have access to it.
> > 
> > 
> > [@Reagan1311](https://github.com/Reagan1311) Hello, why do UMD experiments need to be done on MMSegmentation? Can I not use it and directly train according to the training code of AGD20K?
> 
> Hi, [@MUVGuan](https://github.com/MUVGuan). You can do the UMD experiments based on the training code of AGD20K or any other codebase. I used MMseg because it's convenient to implement other segmentation models.

Got it, thanks for your quick reply!

**Comment by MUVGuan** [comment]
> Hi, here is the UMD one-shot data we used for training: [link](https://drive.google.com/file/d/1MM8n9x68nbRs4xf0zJdgnnq6v1w5ad5m/view?usp=sharing). Following AffCorrs, we have changed the affordance of Ladle from "Contain" to "Scoop". For training and testing, we used mmsegmentation to conduct the experiments.

@Reagan1311 Hello, may I ask if you changed the "contain" of "ladle" to "scoop" in the UMD **testset**?

**Comment by Reagan1311** [comment]
@MUVGuan  Yes, exactly

**Comment by MUVGuan** [comment]
@Reagan1311 
Hello, as for the UMD experiments, do you also use **BCE** loss function to train the network? 
And are there any differences of code when training the network on UMD, since the prediction of AGD is heatmap but it is segmentaion mask for UMD, and the values in heatmap denotes the probabilities but the values in mask denotes the classes. For example, do I need to add a module to change the heatmap prediction of network into a mask firstly when computing the mIoU?

**Comment by Reagan1311** [comment]
@MUVGuan We use BCE or CE loss for UMD experiments. There is no difference during training. For inference, you need to apply a softmax to the output to get segmentation mask to compute the mIoU.

**Comment by MUVGuan** [comment]
> [@MUVGuan](https://github.com/MUVGuan) We use BCE or CE loss for UMD experiments. There is no difference during training. For inference, you need to apply a softmax to the output to get segmentation mask to compute the mIoU.

@Reagan1311 Thanks a lot for your reply! 
1.When computing the CE loss and mIoU, do you include the background?
2.I find that the shape of output is 1×7×224×224, where 7 denotes the number of affordances in UMD, and there is no background. But the GT in UMD is 224*224 and its values belong to {0, 1, 2, 3, 4, 5, 6, 7}. So I'm not sure if I should include background, and the correct way to change the output (1×7×224×224) into a new one included the background (1×8×224×224).
Could you please give some guidances?

**Comment by Reagan1311** [comment]
Hi @MUVGuan. You can compute the CE loss with the background label, but the background should be excluded when computing the mIoU (following the original UMD paper).

**Comment by MUVGuan** [comment]
@Reagan1311 Really thank you for your guidances!
1.As for the visualization of segmentation mask in UMD experiments, can I replace the **Sigmoid** with a **Softmax** (the second line in the following code), and then use **argmax** on **pred** to get mask? 
2.If I can get the mask using the way I said before, how can I get the background (like Fig.6 in your paper), since the shape of pred is 1×7×224×224 and there is no background channel in the second dimension. Maybe using a threshold on pred?
3.When training UMD, are the learning rate and iterations the same as for the AGD20K dataset?

```
attn = (text_features[-1] @ dino_dense.transpose(-2, -1)) * (512 ** -0.5)
attn_out = torch.sigmoid(attn)  # change to softmax?
attn_out = attn_out.reshape(b, -1, h // 14, w // 14)
pred = F.interpolate(attn_out, size=img.shape[-2:], mode="bilinear", align_corners=False)

if self.training:
       assert not label == None, 'Label should be provided during training'
       loss_bce = nn.BCELoss()(pred, label / 255.0)
       loss_dict = {'bce': loss_bce}
       return pred, loss_dict

else:
       if gt_aff is not None:  # tensor([0])
           out = torch.zeros(b, h, w).cuda()
           for b_ in range(b):
                out[b_] = pred[b_, gt_aff[b_]]
           return out
```

**Comment by MUVGuan** [comment]
@Reagan1311 Hello, really sorry to bother you.
In the paper, you said that you use the text prompt template of “somewhere to [affordance]” for MaskCLIP testing, and the mIoU on UMD base setting is 4.25 in Table 2. But I get the mIoU of 3.25 when using only this one template, below are my settings of the experiment, could you please give some guidances? 
1.affordance=['background', 'grasp', 'cut', 'scoop', 'contain', 'support', 'pound', 'wrap-grasp'], ignore_index=0 (i.e. background) when computing mIoU.
2.text prompt template = “somewhere to [affordance]”.
3.transforms: Resize to (224*224), Normalize, and ImageToTensor.

**Comment by Reagan1311** [comment]
@MUVGuan Have you tried PD (prompt denosing) and KS (key smoothing) as stated in their paper? I remember I implement these two methods, and the results get improved a bit.

**Comment by MUVGuan** [comment]
> [@MUVGuan](https://github.com/MUVGuan) Have you tried PD (prompt denosing) and KS (key smoothing) as stated in their paper? I remember I implement these two methods, and the results get improved a bit.

@Reagan1311 Thank you very much for the reply! Let me update the current situation.
1. I found that the Normalize method should use "mean=[123.675, 116.28, 103.53], std=[58.395, 57.12, 57.375]" to replace "mean=[0.485, 0.456, 0.406], std=[0.229, 0.224, 0.225]" in MaskCLIP code, after using it I got mIoU=12.56.
2. Then I used PD and KS with threshold=0.5, and the mIoU changed to 10.28.
Are there any other things to note? Could please give some insights? Thank you so much!

**Comment by Reagan1311** [comment]
Looks good to me, you can follow this setting

**Comment by MUVGuan** [comment]
> Looks good to me, you can follow this setting

But I just used MaskCLIP's official code to experiment on the umd base test set, why is it so much higher than yours In Table2, it's amazing, do you have some insights?

**Comment by Reagan1311** [comment]
@MUVGuan That's strange. Do you also take the background class into accout when computing the mIoU. Also, could you contact me via emails for the following questions, cause It's not appropriate to answer so many questions in one issue.

**Comment by MUVGuan** [comment]
> [@MUVGuan](https://github.com/MUVGuan) That's strange. Do you also take the background class into accout when computing the mIoU. Also, could you contact me via emails for the following questions, cause It's not appropriate to answer so many questions in one issue.

Ok, I will. Thanks a lot for your notification.
